### PR TITLE
QgsMapToolCapture: fix a regression

### LINKED ISSUE #50869: Add Ring/Fill ring tools create a mixed Polygon/CompoundCurve geometry
### What is the bug or the crash?

When using the Add Ring or Fill Ring map tools to create holes in a polygon, the resulting geometry is a combination of Polygon and CompoundCurve with a WKT like 
```
Polygon ((1 2, 4 2, 4 5, 1 5, 1 2),CompoundCurve ((2 4, 2 3, 3 3, 3 4, 2 4)))
```
 instead of the expected
```
Polygon ((1 2, 1 5, 4 5, 4 2, 1 2),(2 4, 2 3, 3 3, 3 4, 2 4))
```

### Steps to reproduce the issue

1. Add a temporary scratch layer of type Polygon
2. Digitize a feature
3. Use Add Ring map tool to punch a hole in your polygon
4. Check the feature's geometry: `print([f.geometry().asWkt() for f in iface.activeLayer().getFeatures()][0])`

### Versions

6cb9a64f8

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

Fix geometries / 0 buffer fixes the geometry to `Polygon ((1 2, 1 5, 4 5, 4 2, 1 2),(2 4, 2 3, 3 3, 3 4, 2 4))`

Achieving the same result with the _Split_ tool gives the expected result `Polygon ((1 2, 1 5, 4 5, 4 2, 1 2),(2 4, 2 3, 3 3, 3 4, 2 4))`

### PR DESCRIPTION
**Pull request by lbartoletti**
## Description

Regression introduced in https://github.com/qgis/QGIS/commit/e2e8a1285022478a257c61e92f130b7ad715a591

Fixes https://github.com/qgis/QGIS/issues/50869

@3nids @uclaros @MattiasSp, can you try this fix?

### CODE REVIEW

File: src/gui/qgsmaptoolcapture.cpp [hunk 1]
```diff
@@ -1381,7 +1381,22 @@ void QgsMapToolCapture::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
     if ( digitizingFinished )
     {
       QgsGeometry g;
-      QgsCurve *curveToAdd = captureCurve()->clone();
+
+      //does compoundcurve contain circular strings?
+      //does provider support circular strings?
+      QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( layer() );
```
Comment (nyalldawson) [inline src/gui/qgsmaptoolcapture.cpp:1387]: You need to check whether layer is nullptr or not a vector layer -- both these situations are possible.

Comment (lbartoletti) [reply src/gui/qgsmaptoolcapture.cpp]: Indeed, fixed

File: src/gui/qgsmaptoolcapture.cpp [hunk 2]
```diff
@@ -1391,6 +1391,29 @@ void QgsMapToolCapture::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
...
+        //does compoundcurve contain circular strings?
+        //does provider support circular strings?
+        QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( layer() );
+        const bool isVectorLayer = ( vlayer != nullptr ) && ( vlayer->type() == Qgis::LayerType::Vector );
+        if ( isVectorLayer )
+        {
+          const bool hasCurvedSegments = captureCurve()->hasCurvedSegments();
+          const bool providerSupportsCurvedSegments = vlayer->dataProvider()->capabilities() & QgsVectorDataProvider::CircularGeometries;
+
+          if ( !isVectorLayer || ( hasCurvedSegments && providerSupportsCurvedSegments ) )
```
Comment (agiudiceandrea) [inline src/gui/qgsmaptoolcapture.cpp:1404]: Isn't `!isVectorLayer` always `false` here?

Comment (agiudiceandrea) [reply src/gui/qgsmaptoolcapture.cpp:1402]: `isVectorLayer` no longer exist now after bf8e32d0a83b562c2e263575af6f75daea852083

Comment (lbartoletti) [reply src/gui/qgsmaptoolcapture.cpp]: yes, it's fixed

**Review by elpaso — APPROVED** [review]
I tested this patch and it fixes the "add ring" part of https://github.com/qgis/QGIS/issues/53698 but it doesn't fix the "fill ring" issue, the hole is created but the new polygon creation fails.

I think we can merge and fix the other issue with a separate PR.

**Comment by agiudiceandrea** [comment]
@elpaso, I've tested this PR and it seems to me both the "Add Ring" and the "Fill Ring" map tools behavior is fixed by it.
@uclaros, may you please double check?

**Comment by agiudiceandrea** [comment]
I think this PR should be backported.

**Pull request merged**